docs/docs/expression_language/cookbook/agents.md

```diff
@@ -0,0 +1,179 @@
+
+```typescript
+const prompt = ChatPromptTemplate.fromMessages([
+  ["system", "You are a helpful assistant"],
+  new MessagesPlaceholder("agent_scratchpad"),
+  ["user", "{input}"],
+]);
+```
+
+After that, we define our structured response schema using zod. This schema defines the structure of the final response from the agent.
+
+```typescript
+const responseSchema = z.object({
+  answer: z.string().describe("The final answer to respond to the user"),
+  sources: z
+    .array(z.string())
+    .describe(
+      "List of page chunks that contain answer to the question. Only include a page chunk if it contains relevant information"
+    ),
+});
+```
+
+Next, we can construct the custom structured output parser.
+
+```typescript
+const structuredOutputParser = (
+  output: AIMessage
+): AgentAction | AgentFinish => {
+  // If no function call is passed, return the output as an instance of `AgentFinish`
+  if (!("function_call" in output.additional_kwargs)) {
+    return { returnValues: { output: output.content }, log: output.content };
+  }
+  // Extract the function call name and arguments
+  const functionCall = output.additional_kwargs.function_call;
+  const name = functionCall?.name as string;
+  const inputs = functionCall?.arguments as string;
+  // Parse the arguments as JSON
+  const jsonInput = JSON.parse(inputs);
+  // If the function call name is `response` then we know it's used our final
+  // response function and can return an instance of `AgentFinish`
+  if (name === "response") {
+    return { returnValues: { inputs }, log: output.content };
+  }
+  // If none of the above are true, the agent is not yet finished and we return
+  // an instance of `AgentAction`
+  return {
+    tool: name,
+    toolInput: jsonInput,
+    log: output.content,
+  };
+};
+```
+
+After this, we can bind our two functions to the LLM, and create a runnable sequence which will be used as the agent.
+
+```typescript
+const llmWithTools = llm.bind({
+  functions: [
+    formatToOpenAIFunction(retrieverTool),
+    {
+      name: "response",
+      description: "Return the response to the user",
+      parameters: zodToJsonSchema(responseSchema),
+    },
+  ],
+});
+/** Create the runnable */
+const runnableAgent = RunnableSequence.from([
+  {
+    input: (i: { input: string }) => i.input,
+    agent_scratchpad: (i: { input: string; steps: Array<AgentStep> }) =>
+      formatForOpenAIFunctions(i.steps),
+  },
+  prompt,
+  llmWithTools,
+  structuredOutputParser,
+]);
+```
+
+Finally, we can create an instance of `AgentExecutor` and run the agent.
+
+```typescript
+const executor = AgentExecutor.fromAgentAndTools({
+  agent: runnableAgent,
+  tools: [retrieverTool],
+});
+/** Call invoke on the agent */
+const res = await executor.invoke({
+  input: "what did the president say about kentaji brown jackson",
+});
+console.log({
+  res,
+});
+```
+
+The output will look something like this
+
+```typescript
+{
+  res: {
+    inputs: "{\n" +
```

Comment:
Why is this `inputs`? And could our output parser parse this into an object?

Reply:
No particular reason, I had just defined that as the output in `structuredOutputParser` (copied from py)
I could spread the response object instead so it would be 
```typescript
{
  res: {
    '  "answer": "The President nominated Circuit Court of Appeals Judge Ketanji Brown Jackson, who is described as one of our nation’s top legal minds and will continue Justice Breyer’s legacy of excellence.",\n' +
    '  "sources": [\n' +
    '    "State of the Union Address"\n' +
    "  ]\n"
  }
}
```

Reply:
Just a bit counterintuitive that the resulting output is `inputs`

Reply:
Ya makes sense, pushed a commit which spreads the return values